internal/transformers/jsx.go

```diff
@@ -0,0 +1,1130 @@
+	acc := &strings.Builder{}
+	initial := true
+	// First non-whitespace character on this line.
+	firstNonWhitespace := 0
+	// Last non-whitespace character on this line.
+	lastNonWhitespace := -1
+	// These initial values are special because the first line is:
+	// firstNonWhitespace = 0 to indicate that we want leading whitsepace,
+	// but lastNonWhitespace = -1 as a special flag to indicate that we *don't* include the line if it's all whitespace.
+	for i := 0; i < len(text); i++ {
+		c, size := utf8.DecodeRuneInString(text[i:])
+		if stringutil.IsLineBreak(c) {
+			// If we've seen any non-whitespace characters on this line, add the 'trim' of the line.
+			// (lastNonWhitespace === -1 is a special flag to detect whether the first line is all whitespace.)
+			if firstNonWhitespace != -1 && lastNonWhitespace != -1 {
+				addLineOfJsxText(acc, text[firstNonWhitespace:lastNonWhitespace], initial)
+				initial = false
+			}
+
+			// Reset firstNonWhitespace for the next line.
+			// Don't bother to reset lastNonWhitespace because we ignore it if firstNonWhitespace = -1.
+			firstNonWhitespace = -1
+		} else if !stringutil.IsWhiteSpaceSingleLine(c) {
+			lastNonWhitespace = i
+			if firstNonWhitespace == -1 {
+				firstNonWhitespace = i
+			}
+		}
+
+		if size > 1 {
+			i += (size - 1)
+		}
+	}
+
+	if firstNonWhitespace != -1 {
+		// Last line had a non-whitespace character. Emit the 'trimLeft', meaning keep trailing whitespace.
+		addLineOfJsxText(acc, text[firstNonWhitespace:], initial)
+	}
+	return acc.String()
+}
+
+func (tx *JSXTransformer) visitJsxExpression(expression *ast.JsxExpression) *ast.Node {
+	e := tx.visitor.Visit(expression.Expression)
+	if expression.DotDotDotToken != nil {
+		return tx.factory.NewSpreadElement(e)
+	}
+	return e
+}
+
+var htmlEntityMatcher = regexp2.MustCompile(`&((#((\d+)|x([\da-fA-F]+)))|(\w+));`, regexp2.ECMAScript)
+
+func htmlEntityReplacer(m regexp2.Match) string {
+	decimal := m.GroupByNumber(3)
+	if decimal != nil {
+		parsed, err := strconv.ParseInt(decimal.Capture.String(), 10, 32)
+		if err == nil {
+			return string(rune(parsed))
+		}
+	}
+	hex := m.GroupByNumber(4)
+	if hex != nil {
+		parsed, err := strconv.ParseInt(hex.Capture.String(), 16, 32)
+		if err == nil {
+			return string(rune(parsed))
+		}
+	}
+	word := m.GroupByNumber(5)
+	if word != nil {
+		// If this is not a valid entity, then just use `match` (replace it with itself, i.e. don't replace)
+		res, ok := entities[word.Capture.String()]
+		if ok {
+			return string(res)
+		}
+	}
+	return m.String()
+}
+
+/**
+* Replace entities like "&nbsp;", "&#123;", and "&#xDEADBEEF;" with the characters they encode.
+* See https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references
+ */
+func decodeEntities(text string) string {
+	res, err := htmlEntityMatcher.ReplaceFunc(text, htmlEntityReplacer, -1, -1)
+	if err != nil {
+		panic(err.Error())
+	}
+	return res
+}
+
+var entities = map[string]rune{
```

Comment:
I would recommend turning this into a switch, since map lookups for stuff like this are not perfect.

Reply:
Map lookups are hash-based but switches do binary searches, no? Unless switches also compile to a hash/perfect hash by the compiler when possible? What's "not perfect" mean here?